### PR TITLE
🌱 Retry CRD detection during startup

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
Retry CRD deletion during start-up, so the controller handle nice race conditions but it goes back to crash loopbacking when the issue is more persistent

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -268,37 +269,46 @@ func main() {
 			return perrors.Wrapf(err, "unable to create remote cluster cache tracker")
 		}
 
-		// Check for non-supervisor VSphereCluster and start controller if found
-		gvr := infrav1.GroupVersion.WithResource(reflect.TypeOf(&infrav1.VSphereCluster{}).Elem().Name())
-		isNonSupervisorCRDLoaded, err := isCRDDeployed(mgr, gvr)
-		if err != nil {
-			return err
+		govimomiGVR := infrav1.GroupVersion.WithResource(reflect.TypeOf(&infrav1.VSphereCluster{}).Elem().Name())
```
Comment (sbueringer) [inline main.go:272]:
```suggestion
		govmomiGVR := infrav1.GroupVersion.WithResource(reflect.TypeOf(&infrav1.VSphereCluster{}).Elem().Name())
```

Comment (sbueringer) [reply main.go:272]: Same below

File: main.go [hunk 2]
```diff
@@ -268,37 +269,46 @@ func main() {
...
+		if err := wait.PollUntilContextTimeout(ctx, 5*time.Second, 30*time.Second, true, func(ctx context.Context) (bool, error) {
+			var errGovimomi, errSupervisor error
+
+			// Check for non-supervisor VSphereCluster and start controller if found
+			isGovimoniCRDLoaded, errGovimomi = isCRDDeployed(mgr, govimomiGVR)
+
+			// Check for supervisor VSphereCluster and start controller if found
+			isSupervisorCRDLoaded, errSupervisor = isCRDDeployed(mgr, supervisorGVR)
+
+			if errGovimomi != nil || errSupervisor != nil {
```
Comment (sbueringer) [inline main.go:285]: Not sure if we should change that logic. Was talking about this with Christian this morning and depending on the exact state of the cluster we can hit issues that we don't cover (e.g. if the apiGroup exists but not the VSphereCluster CRD).

What about changing this to return if one of the (loaded, error) pairs is (true, nil)

File: main.go [hunk 3]
```diff
@@ -268,37 +269,46 @@ func main() {
...
+				return false, nil
+			}
+			return true, nil
+		}); err != nil {
+			return errors.New("failed to detect CRDs")
```
Comment (sbueringer) [inline main.go:290]: Should we use wrap?

File: main.go (hunk 2 above)
Comment (sbueringer) [reply main.go]: > Not sure if we should change that logic

That might have been misleading. I didn't mean that we made a change compared to the previous version. More that we should consider changing the current version on the PR :)

File: main.go [hunk 4]
```diff
@@ -268,37 +269,46 @@ func main() {
...
 
-		// Check for non-supervisor VSphereCluster and start controller if found
-		gvr := infrav1.GroupVersion.WithResource(reflect.TypeOf(&infrav1.VSphereCluster{}).Elem().Name())
-		isNonSupervisorCRDLoaded, err := isCRDDeployed(mgr, gvr)
-		if err != nil {
-			return err
+		govmomiGVR := infrav1.GroupVersion.WithResource(reflect.TypeOf(&infrav1.VSphereCluster{}).Elem().Name())
+		supervisorGVR := vmwarev1.GroupVersion.WithResource(reflect.TypeOf(&vmwarev1.VSphereCluster{}).Elem().Name())
+
+		var isSupervisorCRDLoaded, isGovimoniCRDLoaded bool
```
Comment (chrischdi) [inline main.go:275]:
```suggestion
		var isSupervisorCRDLoaded, isGovmomiCRDLoaded bool
```

File: main.go [hunk 5]
```diff
@@ -268,37 +269,46 @@ func main() {
...
+		if err := wait.PollUntilContextTimeout(ctx, 5*time.Second, 30*time.Second, true, func(_ context.Context) (bool, error) {
+			var errGovimomi, errSupervisor error
+
+			// Check for non-supervisor VSphereCluster and start controller if found
+			isGovimoniCRDLoaded, errGovimomi = isCRDDeployed(mgr, govmomiGVR)
```
Comment (chrischdi) [inline main.go:280]:
```suggestion
			isGovmomiCRDLoaded, errGovmomi = isCRDDeployed(mgr, govmomiGVR)
```

File: main.go [hunk 6]
```diff
@@ -268,37 +269,46 @@ func main() {
...
+				return true, nil
+			}
+			return false, nil
+		}); err != nil {
+			return fmt.Errorf("failed to detect CRDs: %w", err)
+		}
+
+		// Continuing startup does not make sense without having managers added.
+		if !isSupervisorCRDLoaded && !isGovimoniCRDLoaded {
+			return errors.New("neither supervisor nor govimomi CRDs detected")
```
Comment (chrischdi) [inline main.go:295]:
```suggestion
			return errors.New("neither supervisor nor govmomi CRDs detected")
```

**Review by chrischdi — COMMENTED** [review]
This also solves the problem when weird errors get returned for e.g. govmomi mode.

Naming nit :-), as of that LGTM

File: main.go [hunk 7]
```diff
@@ -268,37 +269,46 @@ func main() {
...
 
-		// Check for non-supervisor VSphereCluster and start controller if found
-		gvr := infrav1.GroupVersion.WithResource(reflect.TypeOf(&infrav1.VSphereCluster{}).Elem().Name())
-		isNonSupervisorCRDLoaded, err := isCRDDeployed(mgr, gvr)
-		if err != nil {
-			return err
+		govmomiGVR := infrav1.GroupVersion.WithResource(reflect.TypeOf(&infrav1.VSphereCluster{}).Elem().Name())
+		supervisorGVR := vmwarev1.GroupVersion.WithResource(reflect.TypeOf(&vmwarev1.VSphereCluster{}).Elem().Name())
+
+		var isSupervisorCRDLoaded, isGovmoniCRDLoaded bool
```
Comment (chrischdi) [inline main.go:275]:
```suggestion
		var isSupervisorCRDLoaded, isGovmomiCRDLoaded bool
```

File: main.go [hunk 8]
```diff
@@ -268,37 +269,46 @@ func main() {
...
+			}
+			return false, nil
+		}); err != nil {
+			return fmt.Errorf("failed to detect CRDs: %w", err)
+		}
+
+		// Continuing startup does not make sense without having managers added.
+		if !isSupervisorCRDLoaded && !isGovmoniCRDLoaded {
+			return errors.New("neither supervisor nor govmomi CRDs detected")
 		}
```
Comment (sbueringer) [inline main.go]: This is not possible anymore. One of them has to be true to not end up in an error in l.289

File: main.go [hunk 9]
```diff
@@ -268,37 +269,46 @@ func main() {
...
+
+			// Check for supervisor VSphereCluster and start controller if found
+			isSupervisorCRDLoaded, errSupervisor = isCRDDeployed(mgr, supervisorGVR)
+
+			if (isGovmoniCRDLoaded && errGovmomi == nil) || (isSupervisorCRDLoaded && errSupervisor == nil) {
+				return true, nil
+			}
+			return false, nil
+		}); err != nil {
+			return fmt.Errorf("failed to detect CRDs: %w", err)
```
Comment (sbueringer) [inline main.go:290]: Should we define the errors outside of the poll and add them here with aggregate?

The problem is that we now only print the timeout error but not what actually failed

**Review by chrischdi — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**